File: backend/squad_plan.py
"""Squad-utilization plan (DEC-030).

Pure, pre-flight computation of how a configured squad maps onto a mode's
slots, how many model calls the turn will actually make, and which of those
models are paid. Computed at the squad-resolution chokepoint *before any model
call* (`backend/run_turn.py` for full turns, `TurnService._create_turn_locked`
for stepwise) so the plan can:

  * drive honest participant/quality counts from the *assigned* set rather than
    the raw configured squad (resolves INC-006), and
  * gate execution via the soft-confirm protocol when a turn would be expensive
    or opaque under MCP-driven execution.

This module performs no I/O and makes no model calls. The projected-call
formulas mirror the runners in ``backend/arena.py``; the drift guard in
``tests/unit/test_squad_plan.py`` pins them. ``iterative_schedule`` is the
single source of truth shared with ``run_mode_complex_iterative`` so the plan
and the runner cannot disagree.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import asdict, dataclass
from typing import Dict, List, Optional, Tuple

# --- squad policy -----------------------------------------------------------

QUORUM = "quorum"
REQUIRE_ALL = "require_all"
VALID_POLICIES = (QUORUM, REQUIRE_ALL)


def normalize_policy(value: Optional[str]) -> str:
    """Return a valid policy. ``None``/empty defaults to ``quorum``.

    Raises ``ValueError`` on a non-empty unrecognized value so callers can
    surface a 4xx rather than silently running under a wrong intent.
    """
    if value is None:
        return QUORUM
    candidate = str(value).strip().lower()
    if not candidate:
        return QUORUM
    if candidate not in VALID_POLICIES:
        raise ValueError(
            f"unknown squad_policy {value!r}; expected one of {VALID_POLICIES}"
        )
    return candidate


# --- mode taxonomy ----------------------------------------------------------

# Modes that structurally assign only a subset of the configured squad (or a
# policy-scaled schedule over it). Everything *not* listed here is whole-squad:
# every configured arena model is assigned; there are no reserves.
PARTIAL_SQUAD_MODES = frozenset({"complex_iterative"})

# Provisional soft-confirm threshold (DEC-030): gate a turn for confirmation when
# its exact projected model-call count exceeds this. Paid-call-weighted gating
# (which needs per-role call attribution) is a Phase B refinement; for now the
# paid/free split is surfaced in the plan for the human's cost judgment and the
# gate trigger is call-count based. Phase B wires the threshold to frozen config.
DEFAULT_CALL_THRESHOLD = 12


def is_free_model(model_id: str, free_model_ids: Optional[set] = None) -> bool:
    """A model is free if its id carries the ``:free`` suffix or the catalog
    tagged it free (supplied via ``free_model_ids``)."""
    if ":free" in model_id:
        return True
    return bool(free_model_ids) and model_id in free_model_ids


# --- complex-iterative schedule (shared with the runner) --------------------

_ITERATIVE_ROLES = ("extract", "expand")


def iterative_schedule(
    models: List[str], policy: str
) -> List[Tuple[str, str]]:
    """Ordered ``(role, model)`` steps for Complex Iterative, excluding chair.

    * ``quorum``: the historical behavior — two extract/expand cycles over the
      first two models (``models[0]`` extracts, ``models[1]`` expands). Members
      beyond the first two are reserves.
    * ``require_all``: round-robin every configured model through alternating
      roles for ``max(4, len(models))`` steps so the whole squad participates
      without dropping below the default two-cycle depth. Call count is exact;
      per-model call counts are *not* equal when ``len(models)`` does not divide
      the step budget (e.g. 3 models → 4 steps: m0 appears twice, m1/m2 once).
      ``models_assigned`` still lists each model once; operators should read
      ``role_calls`` / ``projected_calls`` for cost, not assume equal share.
    """
    if not models:
        return []
    if policy == REQUIRE_ALL:
        n = max(4, len(models))
        return [
            (_ITERATIVE_ROLES[i % 2], models[i % len(models)]) for i in range(n)
        ]
    pair = models[:2]
    if len(pair) < 2:
        return []
    return [(_ITERATIVE_ROLES[i % 2], pair[i % 2]) for i in range(4)]


# --- assignment + projected calls ------------------------------------------


def _assigned_and_schedule(
    mode: str, arena_models: List[str], policy: str
) -> Tuple[List[str], List[str], Optional[List[Tuple[str, str]]]]:
    """Split the configured squad into (assigned, reserved) and, for partial
    modes, the ``(role, model)`` schedule computed once from the *full* squad
    so it matches the runner exactly (no second, possibly-divergent computation
    over the deduped assigned set).

    Gate is ``PARTIAL_SQUAD_MODES``: modes not in that set are whole-squad
    (every configured member assigned, no reserves). New partial modes must be
    added there *and* get a schedule implementation in this branch.
    """
    if mode in PARTIAL_SQUAD_MODES:
        # Today only Complex Iterative; its schedule is the shared source of truth.
        schedule = iterative_schedule(arena_models, policy)
        assigned: List[str] = []
        for _role, model in schedule:
            if model not in assigned:
                assigned.append(model)
        reserved = [m for m in arena_models if m not in assigned]
        return assigned, reserved, schedule
    # Whole-squad: every configured member assigned; no reserves.
    return list(arena_models), [], None


def _role_calls(mode: str, assigned: List[str], passes: int) -> Dict[str, int]:
    """Projected *actual model calls* per role (not trace steps) for
    whole-squad modes. Mirrors the runners in ``backend/arena.py``. The notable
    non-identity is Council: its stage-2 rankings are ``M`` parallel calls
    collapsed into a single trace step, so ranking calls equal ``M`` here.
    (Complex Iterative is derived directly from its schedule in
    ``compute_squad_plan``.)
    """
    m = len(assigned)
    if mode == "council":
        return {"answer": m, "ranking": m, "chair": 1}
    if mode == "round_robin":
        return {"draft": passes * m, "chair": 1}
    if mode == "fight":
        return {"answer": m, "critique": m, "defense": m, "chair": 1}
    if mode == "stacks":
        # models[:2] generate, models[2:] critique (or models[:2] if only two).
        generators = min(m, 2)
        critics = (m - 2) if m > 2 else generators
        defenders = generators
        return {
            "answer": generators,
            "merge": 1,
            "critique": critics,
            "judge": 1,
            "defense": defenders,
            "chair": 1,
        }
    if mode == "complex_questioning":
        return {"answer": m, "question_self": m, "brief": 1, "muse": m, "chair": 1}
    # Unknown/baseline: one call per assigned model plus a chair synthesis.
    return {"answer": m, "chair": 1}


# --- plan object ------------------------------------------------------------


@dataclass
class SquadPlan:
    mode: str
    policy: str
    models_assigned: List[str]
    models_reserved: List[str]
    chairman_model: str
    role_calls: Dict[str, int]
    projected_calls: int
    paid_models: List[str]
    free_models: List[str]
    plan_fingerprint: str
    gate_required: bool
    gate_reason: Optional[str] = None

    def to_dict(self) -> Dict:
        return asdict(self)


def _fingerprint(
    mode: str, policy: str, assigned: List[str], chairman: str, projected_calls: int
) -> str:
    payload = json.dumps(
        {
            "mode": mode,
            "policy": policy,
            "assigned": assigned,
            "chairman": chairman,
            "calls": projected_calls,
        },
        separators=(",", ":"),
        sort_keys=False,
    )
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()[:16]


def _evaluate_gate(projected_calls: int, threshold: int) -> Tuple[bool, Optional[str]]:
    if projected_calls > threshold:
        return (
            True,
            f"projected {projected_calls} model calls exceeds threshold {threshold}",
        )
    return False, None


def compute_squad_plan(
    *,
    mode: str,
    arena_models: List[str],
    chairman_model: str,
    policy: Optional[str] = None,
    iterations: Optional[int] = None,
    free_model_ids: Optional[set] = None,
    call_threshold: int = DEFAULT_CALL_THRESHOLD,
) -> SquadPlan:
    """Compute the pre-flight squad plan for a turn.

    ``iterations`` maps to Round Robin passes (``ctx.directives.iterations_override``);
    ignored by other modes. ``free_model_ids`` is an optional catalog-derived set
    of ids tagged free (beyond the ``:free`` suffix heuristic).
    """
    resolved_policy = normalize_policy(policy)
    arena = list(arena_models or [])
    passes = max(1, int(iterations or 1))

    assigned, reserved, schedule = _assigned_and_schedule(mode, arena, resolved_policy)
    if schedule is not None:  # Complex Iterative: derive calls from the schedule.
        role_calls: Dict[str, int] = {}
        for role, _model in schedule:
            role_calls[role] = role_calls.get(role, 0) + 1
        role_calls["chair"] = 1
    else:
        role_calls = _role_calls(mode, assigned, passes)
    projected_calls = sum(role_calls.values())

    # Cost signal: paid/free split across assigned arena models plus the chair.
    cost_models = list(assigned)
    if chairman_model and chairman_model not in cost_models:
        cost_models.append(chairman_model)
    paid_models = [m for m in cost_models if not is_free_model(m, free_model_ids)]
    free_models = [m for m in cost_models if is_free_model(m, free_model_ids)]

    fingerprint = _fingerprint(
        mode, resolved_policy, assigned, chairman_model, projected_calls
    )
    gate_required, gate_reason = _evaluate_gate(projected_calls, call_threshold)

    return SquadPlan(
        mode=mode,
        policy=resolved_policy,
        models_assigned=assigned,
        models_reserved=reserved,
        chairman_model=chairman_model,
        role_calls=role_calls,
        projected_calls=projected_calls,
        paid_models=paid_models,
        free_models=free_models,
        plan_fingerprint=fingerprint,
        gate_required=gate_required,
        gate_reason=gate_reason,
    )


def confirmation_notice(plan: SquadPlan) -> str:
    """Human/agent-facing notice for a gated turn (DEC-030).

    Always instructs the driving model to obtain user approval before
    confirming; it must never self-approve.
    """
    paid = ", ".join(plan.paid_models) if plan.paid_models else "none"
    return (
        f"⚠️ Confirmation required — this turn will make "
        f"{plan.projected_calls} model calls across {len(plan.models_assigned)} "
        f"assigned model(s) plus the chair "
        f"({len(plan.paid_models)} paid, {len(plan.free_models)} free; paid: {paid}). "
        f"Check with the user and get their approval before proceeding — do not "
        f"confirm on their behalf. To proceed after approval, re-invoke with "
        f"confirm=\"{plan.plan_fingerprint}\"."
    )

File: frontend/src/deck/squad-plan.ts
/** Squad-utilization plan helpers (DEC-030 Phase C). */

import type { AssistantMessage } from './types';

export interface SquadPlanView {
  mode: string;
  policy: string;
  models_assigned: string[];
  models_reserved: string[];
  chairman_model: string;
  role_calls: Record<string, number>;
  projected_calls: number;
  projected_paid_calls: number;
  paid_models: string[];
  free_models: string[];
  plan_fingerprint: string;
  gate_required: boolean;
  gate_reason: string | null;
  call_threshold: number;
  paid_call_threshold: number;
}

function stringList(value: unknown): string[] {
  if (!Array.isArray(value)) return [];
  return value.map(String).filter(Boolean);
}

function numberMap(value: unknown): Record<string, number> {
  if (!value || typeof value !== 'object') return {};
  const out: Record<string, number> = {};
  for (const [key, raw] of Object.entries(value as Record<string, unknown>)) {
    const n = Number(raw);
    if (Number.isFinite(n)) out[key] = n;
  }
  return out;
}

/** Parse a plan dict (metadata.squad_plan or top-level plan). */
export function parseSquadPlan(raw: unknown): SquadPlanView | null {
  if (!raw || typeof raw !== 'object') return null;
  const row = raw as Record<string, unknown>;
  const assigned = stringList(row.models_assigned);
  const projected = Number(row.projected_calls);
  // A plan without assigned models is not useful for the roster; treat as absent
  // unless reserves or a fingerprint prove it was intentionally empty.
  if (!assigned.length && !stringList(row.models_reserved).length && !row.plan_fingerprint) {
    return null;
  }
  const projectedPaid = Number(row.projected_paid_calls);
  return {
    mode: String(row.mode || ''),
    policy: String(row.policy || 'quorum'),
    models_assigned: assigned,
    models_reserved: stringList(row.models_reserved),
    chairman_model: String(row.chairman_model || ''),
    role_calls: numberMap(row.role_calls),
    projected_calls: Number.isFinite(projected) ? projected : 0,
    projected_paid_calls: Number.isFinite(projectedPaid) ? projectedPaid : 0,
    paid_models: stringList(row.paid_models),
    free_models: stringList(row.free_models),
    plan_fingerprint: String(row.plan_fingerprint || ''),
    gate_required: Boolean(row.gate_required),
    gate_reason: row.gate_reason != null ? String(row.gate_reason) : null,
    call_threshold: Number(row.call_threshold) || 0,
    paid_call_threshold: Number(row.paid_call_threshold) || 0,
  };
}

/** Resolve the plan from assistant metadata (or a top-level plan field if present). */
export function squadPlanFromMessage(msg: AssistantMessage | undefined | null): SquadPlanView | null {
  if (!msg) return null;
  const meta = msg.metadata || {};
  const fromBlob = parseSquadPlan(meta.squad_plan);
  if (fromBlob) return fromBlob;

  // Reconstruct a minimal plan when only the split fields were persisted.
  const assigned = stringList(meta.models_assigned);
  const reserved = stringList(meta.models_reserved);
  if (!assigned.length && !reserved.length) return null;
  return {
    mode: String(meta.mode || ''),
    policy: String(meta.squad_policy || 'quorum'),
    models_assigned: assigned,
    models_reserved: reserved,
    chairman_model: String(meta.chairman_model || msg.stage3?.model || ''),
    role_calls: {},
    projected_calls: 0,
    projected_paid_calls: 0,
    paid_models: [],
    free_models: [],
    plan_fingerprint: '',
    gate_required: false,
    gate_reason: null,
    call_threshold: 0,
    paid_call_threshold: 0,
  };
}

export interface RosterSplit {
  /** Models expected to produce output (assigned arena set). */
  assigned: string[];
  /** Configured but unscheduled failover pool. */
  reserved: string[];
  /** Full configured squad when known (assigned ∪ reserved, configured order). */
  configured: string[];
  chairman: string;
  policy: string | null;
  hasPlan: boolean;
}

/**
 * Split the turn roster for UI.
 * Prefer the DEC-030 plan; fall back to legacy arena_models for older turns.
 */
export function rosterFromMessage(msg: AssistantMessage | undefined | null): RosterSplit {
  const meta = msg?.metadata || {};
  const plan = squadPlanFromMessage(msg);
  const chairman = String(
    plan?.chairman_model || meta.chairman_model || msg?.stage3?.model || ''
  );
  const configured = stringList(meta.arena_models);

  if (plan) {
    const assigned = [...plan.models_assigned];
    const reserved = [...plan.models_reserved];
    // Preserve configured order when available; otherwise assigned then reserved.
    const ordered =
      configured.length > 0
        ? configured.filter((m) => assigned.includes(m) || reserved.includes(m))
        : [...assigned, ...reserved];
    // Include any assigned/reserved not present in arena_models (defensive).
    for (const m of [...assigned, ...reserved]) {
      if (!ordered.includes(m)) ordered.push(m);
    }
    return {
      assigned,
      reserved,
      configured: ordered.length ? ordered : configured,
      chairman,
      policy: plan.policy,
      hasPlan: true,
    };
  }

  return {
    assigned: configured,
    reserved: [],
    configured,
    chairman,
    policy: meta.squad_policy != null ? String(meta.squad_policy) : null,
    hasPlan: false,
  };
}

export function formatRoleCalls(roleCalls: Record<string, number>): string {
  const entries = Object.entries(roleCalls);
  if (!entries.length) return '';
  return entries
    .map(([role, count]) => `${role.replace(/_/g, ' ')}×${count}`)
    .join(' · ');
}
